### ISSUE TITLE
The botclip shader/surfaceparm has no effect.

### ISSUE THREAD

**Issue by Viech** [Unvanquished/Unvanquished#1443]
What the title says. Bots can (and will) go through it. Either make it work, or if that is not possible, remove it from [here](https://github.com/UnvanquishedAssets/tex-common_src.dpkdir/blob/master/scripts/common.shader#L47).

A comment from 2014 saying that it doesn't work is no excuse to keep it around disfunctional. :stuck_out_tongue:

**Comment by slipher** [comment]
What would be the expected semantics?

Put it on the ground and expect bots not to walk on it? Or use it like a wall?

Should it physically block bots or should they try to avoid it when navigating, but still can pass over/through?

**Comment by Viech** [comment]
Since it is a clip, it should phiscally block bots. Like playerclip, but just for bots. You can use it to seal off niches where they get stuck or build them invisible ramps to get out, for instance.

**Comment by slipher** [comment]
That would have some interesting effects... like if you build a telenode intersecting the surface, then the bot will be stuck.

**Comment by Viech** [comment]
The very same can be said about `playerclip`, which is omnipresent. Also, I think (but I'm not sure) that you can move out of a clip brush, just not into.

**Comment by slipher** [comment]
I would consider it a bug if you can build somewhere that players can't go.

The ramp idea sounds nice for something like the very shallow pits bots get stuck in on Parpax. But it would look buggy in-game to have invisible walls for bots... say you're trying to pounce someone off a ledge, but then they're stopped by an invisible wall.

Another possibility is something to put on the floor which would inhibit navmesh generation. This could be used to work around the problem of small areas of navmesh which are not connected to the main component, by applying it to places where the mesh generator is discovered to do that.

You could also try to use this method to prevent bots from walking off a ledge, by putting some on the floor near the edge. I don't expect it would be that effective though.

**Comment by Viech** [comment]
> I would consider it a bug if you can build somewhere that players can't go.

You misunderstand how player clips are normally used. If you have a detailed wall with a little depth, say pipes going across it or with a shallow cavity, then you still want dretches to wallwalk it without joggle and you want the marauder to climb it without getting blocked by some detail. So you clip it flat. No one has ever noticed that the bounding box of aliens walking that wall is a few units off. However, when you build an egg or booster on that wall, you don't want it to float by a few units because one would definitely notice that. That's why playerclip does not affect buildings. Its standard use is not to seal off whole areas, but to flatten a surface.

> The ramp idea sounds nice for something like the very shallow pits bots get stuck in on Parpax. But it would look buggy in-game to have invisible walls for bots... say you're trying to pounce someone off a ledge, but then they're stopped by an invisible wall.

Agreed, although the downside is often worse than this little oddity. For instance, on the new parpax version human bots must use the D-stairs to go up to the alien base. And they rarely make it without falling from the top bridge as they start strafing when they see an enemy. I'd rather have invisible walls than this. (Also, when \_ds\_ was working on bots, at some point he had planned to simply confine them to the navmesh boundaries, i.e. have invisible walls virtually everywhere. I don't blame him any more, this would actually solve almost all bot problems we are having for years.)

> Another possibility is something to put on the floor which would inhibit navmesh generation.

Doesn't prevent bots from getting stuck there at all. In fact, them getting stuck implies that there is a gap in the navmesh, so even if they end up on a navmesh island, then we still know that they never intended to go there. They only see the connected navmesh component that they are on. Falling down is the same; a smaller navmesh won't prevent it because they ignore the navmesh in the first place.

**Comment by slipher** [comment]
Sure, a few units of playerclip is fine; I mean it would be a bug if you could stick a whole building in there.

I could be wrong but I think bots usually do OK when off mesh, managing to walk back to it. The point is to prevent navmesh islands from existing (since they trap bots), not to prevent bots from going there.

**Comment by necessarily-equal** [comment]
So, if we want someone to implement the botclip, we need to define how it should work.

I have an usecase for botclip: I would like to prevent bots to attack the enemy's base on a tutorial map. Let's say a (alien) bot clip that would protect the human base. Let's review the two proposals in that light.

If we implement the botclip so that these zones are excluded from the navmesh, it works quite well: bots will not notice someone or some building outside of the navmesh, but will pursue someone fleeing in a no-bot zone. But maybe @bmorel has changes in mind that will break this current behaviour.

For the hard-wall botclip. I think it would works well enough too (I couldn't test it), but bots will never pursue someone in a botclip zone.

**Comment by necessarily-equal** [comment]
The other point I see in a botclip that has been suggested in this thread is with a hard-wall botclip. It can be used to build "bridges" on the ground to fix the navmesh. And this could be extended to other places as well. For example, we can make a botclip "ramp" on the boulders on chasm to connect the navmeshes of the "elevated island".

Maybe I missed something, but I don't think the other usages are significant enough to be worth it.

**Comment by illwieckz** [comment]
The root cause of botclip not working is that we never attributed it a surface flag so that data get baked in bsp and then didn't implemented that clip in game. The reason why we never attributed a surface flag for bot clip is because surface flags are among the rarest resources we have. We only have one bitfield for surfaceflags, once every bits of the bitfield, we can't add one more flag without forking the BSP format and cutting down the Unvanquished branch from Q3 BSP tree with two decades of produced maps.

Some commercial games did that, while Q3 derivatives usually have two bit fields (one for surface parms, and one for content parms), some Raven games used two bit fields for surface parms and one bit field for content field. It allowed them to have many “something-step” sound effects like grass-step, snow-step or things like that. But for now the engine has no way to support multiple BSP, even not the XreaL BSP which had fixes for other shortcomings.

Though, we are not out of solutions, we must just thing thrice when doing things. Tremulous wasted a lot of surface flags and content flags, as seen in that comment: https://github.com/Unvanquished/Unvanquished/issues/1373#issuecomment-821862567 the `noalienbuild` and `nohumanbuild` flags are almost unused, even sometime the flag is not set while the material is used… We can safely drop `noalienbuild` and `nohumanbuild` flags in both contentparm and surfaceparm bit fields.

Note: I don't know if that can apply to `botclip` but since in Unvanquished we expect all materials to have a `.shader` file and the Dæmon engine know how to read the `surfaceparm` fields, once the q3map2 map compiler keeps the surface in the BSP in a way the engine processes it, we may apply `surfaceparm` flags in a in-engine bitfield that is not stored in BSP file at all. This may be enough for special surface step parm, and maybe botclip as well…

**Comment by bmorel** [comment]
TL;DR: I don't think this feature is useful. The reason why bots fall is that when they fight they ignore the navmesh. Adding invisible walls is working around instead of fixing the problem.
Also, it is already possible to add junctions between navmeshes, which works as long as bots can "slide": jumpers, teleports, buggy navmesh generations, from up to down, and even some elevators can be used by bots when doing that.

> > The ramp idea sounds nice for something like the very shallow pits bots get stuck in on Parpax. But it would look buggy in-game to have invisible walls for bots... say you're trying to pounce someone off a ledge, but then they're stopped by an invisible wall.
> 
> Agreed, although the downside is often worse than this little oddity. For instance, on the new parpax version human bots must use the D-stairs to go up to the alien base. And they rarely make it without falling from the top bridge as they start strafing when they see an enemy. I'd rather have invisible walls than this. (Also, when _ds_ was working on bots, at some point he had planned to simply confine them to the navmesh boundaries, i.e. have invisible walls virtually everywhere. I don't blame him any more, this would actually solve almost all bot problems we are having for years.)

Here, the solution you built is way better: put actual physical ramps, which will help both bots, but also players, to not fall. Considering the game is built to be played by human beings, I think it's a better solution :)

Overall, I do not like the idea of having bot markers to force them to do (or to not do) something. Note however, that I am not against giving them markers giving them hints about an action being possible: I realize that there might not be other choice to make them know that there is a passage at some place.
This can already be done, since it is possible to (manually) connect navmeshes with either one-way or two-way passages, but bots still currently only know how to fall, they don't really interpret those markers.
They are, indeed, really dumb, but their dumbness still can allow them to use all elevators in new parpax, since you bring this map in the discussion (basically, do "/addcon start oneway" at default position trigger, and "/addcon end" when the elevator leads to in "/navedit enable human_naked". Then, since you should be lazy, duplicate the file are rename it for each class which can reach the trigger: done.).
Of course, that trick does not allows them to use (all but one) elevators in spacetracks: they're too complex, and bots can't activate the triggers "by error".

> > Another possibility is something to put on the floor which would inhibit navmesh generation.
> 
> Doesn't prevent bots from getting stuck there at all. In fact, them getting stuck implies that there is a gap in the navmesh, so even if they end up on a navmesh island, then we still know that they never intended to go there. They only see the connected navmesh component that they are on. Falling down is the same; a smaller navmesh won't prevent it because they ignore the navmesh in the first place.

Bots actually only ignore the navmesh when they are fighting. This distinction is important, imo, because it means this can be fixed with currently working features.
However, I don't say it will be easy.

> I could be wrong but I think bots usually do OK when off mesh, managing to walk back to it.

Sadly, this is wrong. There are lot of cases in which they suicide or can't move out of those.
There are 2 solutions here, and I think ideally, a mix of both shuold be applied.
As you stated, avoiding navmesh islands is one of them. I think this should be done by mappers when said islands give nothing to gameplay except annoying even players, like the ones previous parpax iteration had. Current version fixed them, and I think it's not only better for bots, but also for humans.

> I have an usecase for botclip: I would like to prevent bots to attack the enemy's base on a tutorial map. Let's say a (alien) bot clip that would protect the human base.

I think I already said taht on IRC, but I really think you should simply write specific behavior tree which only makes the "bad bots" to roam in radius instead or roam/rush. To prevent bots chasing too much, you can simply check distance with enemy main buildable. I agree that, on this, we need a way to check distance with either the closest friendly or closest enemy buildables, but that seems easier to do than a new clipping. I can think of several other things missing in the BT's API, which could be easy to implement and serve that usecase (like a "in-range" check).
Of course, there are more hackish solutions...  of all solutions for the use case you're talking about, the use of a botclip is the one requiring the most work and also not the cleanest one.

> maybe @bmorel has changes in mind that will break this current behaviour.

I don't, but that's not relevant.

> It can be used to build "bridges" on the ground to fix the navmesh.

It is already possible to manually add connections, which is a less ugly solution, since it does not requires physically patching the map. Well, admittedly, that's only when bots can "slide" or trigger something moving them. What is really needed here, is to implement a way to make them know what to do to pass through a particular connection: this means adding meta data to those, and likely integrating that notion into either netradiant, unvanquished or both (considering the fact unvanquished already have it, but just lacks the meta data stuff).

**Comment by necessarily-equal** [comment]
I still think it makes sense. I've implemented something like Gireen suggested, and went as far as adding the `hysteresis` BT node¹ for it, and to be honest, it's just not as good as removing the navmesh. (At least with the current behaviour that bot have, when they ignore the navmesh when in sight, which I quite like btw.)

The investigation would also allow removing the weird navmesh generated everywhere², including on the sky and on deadly places, which may become a problem if we want to generate navmesh connexions automatically.

¹: see the open PR that introduces it
²: there is an open issue by illwieckz about that

Le 1 février 2022 16:20:22 GMT+01:00, "Morel Bérenger" ***@***.***> a écrit :

**Comment by Viech** [comment]
Just to clarify, a _clip_ is always a hard wall that stops certain traces through the game world. So for `botclip` to be a proper clip, it would need to physically prevent bots from going through it (including involuntarily). Thus it should also be considered a solid wall by the navmesh generator.

**Comment by bmorel** [comment]
> At least with the current behaviour that bot have, when they ignore the navmesh when in sight, which I quite like btw

They only ignore the navmesh if you do `action fight`. And if you want to prevent them from fighting too far, you can use the `concurrent` node.

Example:

```
selector {
  concurrent {
    action fight
    condition distanceTo( E_ENEMYBUILDING ) > 1000
  }
  roamInRadius( E_FRIENDLYBUILDING, 500 )
}
```

In this snippet, a bot should fight, but will keep checking for the distance to any enemy building. If too close, it will run away, even if still figthing.

**Comment by slipher** [comment]
This discussion is getting super unrelated to the issue, but if the goal is to stop fighting when the distance is over 1000, can't that be done by
```
condition distanceTo( E_ENEMYBUILDING ) > 1000 {
    action fight
}
```

**Comment by bmorel** [comment]
> This discussion is getting super unrelated to the issue,

True. Also, for the need expressed, a mapper can place an entity which can block navmesh. In addition with BTs, this should work, *in current state*, and thus seems much better to me than making the map format more unvanquished specific than it already is (and wasting one of the rare remaining bits).

To have bot specific invisible walls will lead to bots running against the "wall" and will be quite weird

**Comment by Viech** [comment]
> To have bot specific invisible walls will lead to bots running against the "wall" and will be quite weird

Not if the navmesh generater treats `botclip` exactly the same as a `fullclip`: Like any other solid wall.

**Comment by bmorel** [comment]
Won't change anything since currently fighting bots rely on traces and ignore navmesh.

**Comment by Viech** [comment]
Oh I see what you mean. But don't think about `botclip` as an actual wall: If it was functional, it would be used e.g. to seal off niches that require specific movement to get out of, not unlike `playerclip` is used to make wallwalking on detailed walls smooth. For instance there is this one niche with a pipe on parpax that bots used to get stuck in a lot. With botclip I could build small ramps so that they could just walk out of it while humans need to jump. My workaround was to use playerclip, which is not a bad solution for this scenario, but you will see that there are certain niches where it is more natural to require players to use movement other than walking to cross them (but which is too hard to do for bots).

**Comment by bmorel** [comment]
If you're talking about this pipe:

![2022-02-01-204427_1362x748_scrot](https://user-images.githubusercontent.com/1316300/152039907-26a36919-3c30-4d7d-a0f6-8ceb8e6fe63c.png)

Then there's an easy solution: just add an offmesh connection. It seems a lot less troubles to me than adding bot specific geometry.
Also, I have good hopes to be able to generate navmeshes with some auto-jump code in them, as a start. In the end, I've found people explaining it's possible to get more advanced stuff, including automatic off-mesh connections, and I clearly hope to get them.
The pipe I screenshotted may very well be reachable with the "lesser auto-jump" feature I described, when I'll get rid of the other problems it brings (running toward the middle pillar and jumping allows to go on pillar, and that place is connected to the pipe).

If you're not talking about this pipe, I'm curious to know about which one you're talking.

**Comment by Viech** [comment]
Nope, this one:

![unvanquished_2022-02-01_205536_000](https://user-images.githubusercontent.com/3780206/152041671-2da6aaa4-c0f4-42fb-ac9d-b728dcfcd567.jpg)

> Then there's an easy solution: just add an offmesh connection.

IIRC adding offmesh connections is the precise opposite of easy. It would be easy if it could be done within radiant as part of your normal mapping workflow. Editing the compiled navmeshes (there is technically one for every class) with a set of special commands within unvanquished and redoing this everytime your map changes while keeping track of this with every release is prohibitibely hard.

**Comment by bmorel** [comment]
Oh you meant previous iteration!
Well, for that case, the lesser auto-jump feature I have a PoC for is perfect, actually. Same for this place in nexus6:

![2022-02-01-210542_1362x748_scrot](https://user-images.githubusercontent.com/1316300/152042918-0c1096a3-043c-44ef-98c3-9d6de38e7c4e.png)

> This could indeed help with a lot of problems that botclip would be useful for.

Exactly. The auto-jump feature I'm working on is to fix that kind of annoyances. The idea is, tbh, pretty simple: instead of using the STEPSIZE value as what bots can climb, use STEPSIZE+jumpHeight, and rely on traces to detect that they need to jump (they already do that anyway, notably to jump above entities like buildings)

As I said, I still have some problems with it, but I have verified that it do fix the problem of the crouching and what I call "lesser auto-jump", that is, jumps which are smaller than the character's class.

**Comment by Viech** [comment]
Keep in mind though that having a _lesser_ auto-jump implies the existence of a _legendary_ auto-jump, catapulting the bot from one map to another. :rocket: 

Alright; back to topic below this line:

---

`botclip`: Either make work as advertised or remove (I don't have a strong opinion which one).

**Comment by illwieckz** [comment]
If I'm right I have somewhere a working implementation of `botclip` (both in daemonmap and in game), I'm just super busy right now.

**Comment by bmorel** [comment]
If that's not used in any map, uses precious resources and have no point except that bots' pathfinding in current state is more or less a PoC, I think it would be better to not integrate it, really.
For the ** legendary jump** thing and otherwise big connections, if it was simply possible to add dummy entities with a special type, it could be possible to mappers to add them in maps themselves.

For such a feature (since I'm at it, let's describe what I suppose it would need...) those dummy entities would need:

* to be linked together (links have a start and an end);
* to be affected to abilities, which are currently:
  * takes fall damages
  * can use ladder
  * wall climbing
  * wall jumping
  * wall running
  * can zoom (not useful here though)
  * fov warps (no idea what's this, not useful here though)
  * aliensense (this is not really useful for that need, and I wonder if it is really useful at all, and if yes, why no radar in there?)
  * slider (no idea what this is: all humans except bsuit have it... is that the thing which makes humans like they walk on soapy floors?)
  * can fly (DO NOT EXISTS YET)
  * can pounce (DO NOT EXISTS YET)

This way, unvanquished itself would be able to create off-mesh connections (navlinks) automatically when loading navmeshes, which is something I'll need to implement anyway for elevators and doors. Of course, I'll also need to implement AI support for those, but that's on my short term todo list.

I think implementing proper bot navigation would make it useless and even hurtful to have that "AI clip" feature.

**Comment by Viech** [comment]
The legendary jump thing was a joke (w.r.t. _lesser/greater/legendary_ keywords in RPGs) and at this point you are misusing this issue. Everything apart from your first and last sentence should not be discussed here further. This issue is about a feature being in the game and not doing anything, to the confusion of mappers.

With respect to `botclip` being "hurtful", I don't see this at all. All other clips have perfectly valid uses. Even with very good navmesh generation and bot navigation, one should be able to build map geometry that is challenging to humans and near impossible to navigate for bots without some duct tape in the form of clips. Though, again, I'm fine with `botclip` being removed. It is useful but not strictly necessary.

**Comment by bmorel** [comment]
> The legendary jump thing was a joke

I know I got it :)

> With respect to botclip being "hurtful", I don't see this at all.

If this "feature" is abused, it will make bots look like buggy, and if used systematically in places like like one you pointed in old parpax, it will either make bot debugging harder because they'll have invisible obstacles (yes, this debug occurs in game, which I bet have no way to see those by default, and no I don't want to use $radianteditor to debug bots, it's hard enough as it is) and make things less reproducible.

**Comment by Viech** [comment]
That's a good point, it would be the only clip hard to experience/reproduce as a human tester.

**Comment by illwieckz** [comment]
Though this may be doable to do a cheat cvar to make non-bots be affected by botclip.

**Comment by bmorel** [comment]
bot pathfinding debugging is mostly about enabling navedit and watching at them in the hope they get into the correct situation (I don't know how to do it a better way yet), looking at the the small line which shows where they want to go.
Usually trying to pause/timescale to make it more observable, and sometimes trying to rely on gdb, too.
Even if players can suffer the botclip, this won't help much, not as long as there is not a *proper* and easy way to test bot moves. Would start by a minimal GUI to do it. And I think that'd be required before such botclip "feature". At current status, if I know botclip is used in a map, I'll simply consider the map as inherently buggy and not-fixable. And since this won't be advertised, I'll likely loose a lot of time to just notice that fact.

**Comment by illwieckz** [comment]
See https://github.com/DaemonEngine/daemonmap/pull/14 for an example of botclip implementation on daemonmap side.

Now I believe I didn't have a game-side implementation because playerclips are done the contentparm way, and botclip is currently a surfaceparm (and I would prefer to not add a contentparm).

On the other hand, we may also reuse monsterclip contentparm, which is already defined and unused. If I'm right monsterclip was for scripted ennemies, that's the closest thing to a bot.

**Comment by bmorel** [comment]
In my opinion, mappers should (ideally) *never* have to bother about bot movements.
Bots should act as normal players. If a server owners want training bots, then behavior trees should be the target.
I don't see any point in such feature. Especially now taht I've skimmed the 2001 thesis about Q3's bots, which comforted me in many of my current (as of july 2022) PoV.

Hence, this non-working feature should be removed.

**Comment by bmorel** [comment]
@illwieckz also... do the map editors actually support that anti-feature, or is it just unvanquished or daemon?
